### PR TITLE
Bluetooth: host: Fix DB hash calculation

### LINKED ISSUE #39131: GATT DB hash calculation is wrong on characteristic declarations using 128-bit UUIDs.
**Describe the bug**
In the function https://github.com/zephyrproject-rtos/zephyr/blob/ef0d955758e549f39d97b07089dedc4bdd05bc0e/subsys/bluetooth/host/gatt.c#L609-L682

when hashing attributes for the UUID cases:
BT_UUID_GATT_PRIMARY_VAL
BT_UUID_GATT_SECONDARY_VAL
BT_UUID_GATT_INCLUDE_VAL
BT_UUID_GATT_CHRC_VAL
BT_UUID_GATT_CEP_VAL

where (Attribute Handle || Attribute Type || Attribute Value) is to be hashed, the worst case size of attribute value is 19:

Characteristic declaration (BT_UUID_GATT_CHRC_VAL)
Attribute Value:
* Characteristic Properties: 1 octets
* Characteristic Value Handle: 2 octets
* Characteristic UUID: 2 or 16 octets

The data array used to store the attribute value is only 16 bytes long and therefor the last 3 bytes of the characteristic UUID is not included in the hash.

**Expected behavior**
Proper GATT DB hash generation.

**Impact**
Under certain conditions changes in UUID in the GATT DB will not be detected by a client, leading to errors accessing attributes

### PR DESCRIPTION
**Pull request by thoh-ot**
When hashing attributes for the UUID cases:
    BT_UUID_GATT_PRIMARY_VAL
    BT_UUID_GATT_SECONDARY_VAL
    BT_UUID_GATT_INCLUDE_VAL
    BT_UUID_GATT_CHRC_VAL
    BT_UUID_GATT_CEP_VAL

where (handle || UUID || value) is to be hashed, the worst case size of
attribute value is 19:

Characteristic declaration (BT_UUID_GATT_CHRC_VAL)
Attribute Value:
* Characteristic Properties: 1 octets
* Characteristic Value Handle: 2 octets
* Characteristic UUID: 2 or 16 octets

Update the data array to accomodate for the worst case size.

Fixes #39131

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

### CODE REVIEW

**Comment by thoh-ot** [comment]
There's an issue in the EDTT GATT test where the contents of 'GATT Database.xml' no longer matches, as the hash calculation now computes a different number for the 'Database Hash' characteristic.

**Review by alwa-nordic — CHANGES_REQUESTED** [review]
Changes have been agreed on in conversation.

File: subsys/bluetooth/host/gatt.c [hunk 1]
```diff
@@ -611,7 +611,7 @@ static uint8_t gen_hash_m(const struct bt_gatt_attr *attr, uint16_t handle,
 {
 	struct gen_hash_state *state = user_data;
 	struct bt_uuid_16 *u16;
-	uint8_t data[16];
+	uint8_t data[19];
```
[the remark this answers is not in the document]
Comment (carlescufi) [reply subsys/bluetooth/host/gatt.c:614]: @thoh-ot can you please create the complete union and use that?

**Comment by thoh-ot** [comment]
As mentioned in the issue some of the GATT EDTT tests are failing with the DB hash fix, because there's an generated XML file with expected GATT DB values in it, and it no longer matches. 

As the consultant who implemented that system is no longer in the company, I have done some archaeological digging in our version control system, and we found the following:

In a PTS release there is an Excel sheet, with multiple reference GATT databases, that are to be used during PTS testing, and also GATT EDTT testing as they mimic the PTS tests.

This Excel sheet  and been modified and Visual Basic for Application code added for generating a XML file for the GATT EDTT python test code, and .c & .h files for the corresponding GATT DB in the DUT bsim code.

I tried to get the VBA code running and running, but there seems to be some hardcoded paths and from the look of it, it would generate code for older types and apis.

Which PTS release the Excel sheet originally came from i don't know, but i can see that it doesn't match the newest PTS release, where things having added and moved around in reference databases to enforce certain behavior.

The issue is how do we wan't to approach this now, would try to revive the generation of XML file, c and h files or would we just hardcode the failing GATT DB hashes?

**Comment by carlescufi** [comment]
> As mentioned in the issue some of the GATT EDTT tests are failing with the DB hash fix, because there's an generated XML file with expected GATT DB values in it, and it no longer matches.
> 
> As the consultant who implemented that system is no longer in the company, I have done some archaeological digging in our version control system, and we found the following:
> 
> In a PTS release there is an Excel sheet, with multiple reference GATT databases, that are to be used during PTS testing, and also GATT EDTT testing as they mimic the PTS tests.
> 
> This Excel sheet and been modified and Visual Basic for Application code added for generating a XML file for the GATT EDTT python test code, and .c & .h files for the corresponding GATT DB in the DUT bsim code.
> 
> I tried to get the VBA code running and running, but there seems to be some hardcoded paths and from the look of it, it would generate code for older types and apis.
> 
> Which PTS release the Excel sheet originally came from i don't know, but i can see that it doesn't match the newest PTS release, where things having added and moved around in reference databases to enforce certain behavior.
> 
> The issue is how do we wan't to approach this now, would try to revive the generation of XML file, c and h files or would we just hardcode the failing GATT DB hashes?

I am not sure how this is related to the code in this PR however, and to the discussion we are having regarding the structs and unions?

**Comment by thoh-ot** [comment]
It's the reason for why the CI is failing, no matter if we use unions, structs or just increase the array length, the CI will never pass. 

Fixing the GATT DB hash calculation will render the generated files located in EDTT and tests/bluetooth/bsim_bt/edtt_ble_test_app/gatt_test_app/src/gatt invalid and thus the CI will fail.

**Comment by carlescufi** [comment]
> It's the reason for why the CI is failing, no matter if we use unions, structs or just increase the array length, the CI will never pass.
> 
> Fixing the GATT DB hash calculation will render the generated files located in EDTT and tests/bluetooth/bsim_bt/edtt_ble_test_app/gatt_test_app/src/gatt invalid and thus the CI will fail.

Oh, I see. Thanks for the explanation @thoh-ot

**Comment by carlescufi** [comment]
@thoh-ot even without fixing the .xml file and/or regenerating it, couldn't we merge the fix so at least we can test it with PTS and 128-bit UUIDs?

File: subsys/bluetooth/host/gatt.c [hunk 2]
```diff
@@ -606,12 +606,48 @@ struct gen_hash_state {
 	int err;
 };
 
+union hash_attr_value {
+	/* Bluetooth Core Specification Version 5.3 | Vol 3, Part G
+	 * Table 3.1: Service declaration
+	 */
+	union {
+		uint16_t uuid16;
+		uint8_t  uuid128[16];
```
Comment (Thalley) [inline subsys/bluetooth/host/gatt.c:615]:
```suggestion
		uint8_t  uuid128[BT_UUID_SIZE_128];
```

File: subsys/bluetooth/host/gatt.c [hunk 3]
```diff
@@ -606,12 +606,48 @@ struct gen_hash_state {
...
+	} __packed inc;
+	/* Bluetooth Core Specification Version 5.3 | Vol 3, Part G
+	 * Table 3.3: Characteristic declaration
+	 */
+	struct {
+		uint8_t properties;
+		uint16_t value_handle;
+		union {
+			uint16_t uuid16;
+			uint8_t  uuid128[16];
```
Comment (Thalley) [inline subsys/bluetooth/host/gatt.c:634]:
```suggestion
			uint8_t  uuid128[BT_UUID_SIZE_128];
```

**Comment by thoh-ot** [comment]
@carlescufi I have "shelfed" the generation tool, and instead updated the "Reference Database Hash" in the EDTT GATT test with the correct value and also added an implementation of the Bluetooth SIG GATT test "GATT/SR/GAS/BV-02-C" however the CI doesn't pull in the Python module update (Pycryptodome) needed for the EDTT GATT test change to fix the failing tests, any idea on how to proceed?

**Comment by thoh-ot** [comment]
The Python module update requires an update to the CI docker image but to speed this pr along, i have decided to split the task in two and only focus on (1) in this PR.
(1) Fix Host GATT DB hash calculation and Update EDTT with corect DB hash
(2) Introduce new EDTT GATT test (with new Python module)

File: subsys/bluetooth/host/gatt.c [hunk 4]
```diff
@@ -606,12 +606,48 @@ struct gen_hash_state {
...
+	} __packed service;
+	/* Bluetooth Core Specification Version 5.3 | Vol 3, Part G
+	 * Table 3.2: Include declaration
+	 */
+	struct {
+		uint16_t attribute_handle;
+		uint16_t end_group_handle;
+		uint16_t uuid16;
+
+	} __packed inc;
```
Comment (Thalley) [inline subsys/bluetooth/host/gatt.c:625]:
```suggestion
		uint16_t uuid16;
	} __packed inc;
```

Comment (thoh-ot) [reply subsys/bluetooth/host/gatt.c:625]: Ack

**Review by Thalley — APPROVED** [review]
LGTM

**Comment by thoh-ot** [comment]
Waiting for the EDTT PR https://github.com/EDTTool/EDTT/pull/50 to be merged and then mirrored over into https://github.com/zephyrproject-rtos/edtt, so i can update the edtt west sha to point to the correct sha.

**Comment by alwa-nordic** [comment]
@thoh-ot The EDDT PR is merges now. Should we move this out of draft status?

**Review by wopu-ot — APPROVED** [review]
LGTM

**Comment by thoh-ot** [comment]
@alwa-nordic have your 'request changes' been addressed?

File: subsys/bluetooth/host/gatt.c [hunk 5]
```diff
@@ -606,12 +606,47 @@ struct gen_hash_state {
...
+	/* Bluetooth Core Specification Version 5.3 | Vol 3, Part G
+	 * Table 3.3: Characteristic declaration
+	 */
+	struct {
+		uint8_t properties;
+		uint16_t value_handle;
+		union {
+			uint16_t uuid16;
+			uint8_t  uuid128[BT_UUID_SIZE_128];
+		};
```
Comment (alwa-nordic) [inline subsys/bluetooth/host/gatt.c:634]: Missing __packed here.

**Review by jhedberg — APPROVED** [review]

File: subsys/bluetooth/host/gatt.c (hunk 5 above)
Comment (thoh-ot) [reply subsys/bluetooth/host/gatt.c]: Fixed

**Comment by thoh-ot** [comment]
> Changes have been agreed on in conversation.

@alwa-nordic Is everything good now?

**Comment by jhedberg** [comment]
@alwa-nordic please take another look here. Your review is blocking merging.

**Review by alwa-nordic — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@thoh-ot thanks for following this up, this look good now.

**Pull request merged**